### ISSUE TITLE
ComposerRepository::getProviderNames has different behavior in case providers are provided as partial packages

### ISSUE THREAD

**Issue by boesing** [composer/composer#8526]
Hey there,

I am actually migrating some projects using private packages to laminas.
Laminas is using the [laminas/laminas-dependency-plugin](https://github.com/laminas/laminas-dependency-plugin) to manipulate installation operations (they change the package inside the operation to avoid installing one of the old packages).

In laminas/laminas-dependency-plugin#6 I've reported a bug, that our projects do not properly migrate due to the `{"packagist.org": false}` entry in our `repositories` property.

After some debugging with composer, I found out, that the provider names are not properly returned for lazy packages. Private packagist creates a `packages.json` with `packages` and `providers-lazy-url` which `packagist.org` does not have. 

I've created a possible fix in https://github.com/composer/composer/pull/8516

Would love to get any feedback in the PR.

**Comment by Seldaek** [comment]
Do you have the laminas packages present in your private packagist instance or not? If yes I assume the plugin should work correctly, but I am not 100% sure I understand the implications of the hackery they are doing here tbh.

In any case your patch seems harmless enough that I am fine merging it, but I don't really understand why it fixes things, nor what the problem is exactly that you experienced.

**Comment by boesing** [comment]
Yes, they are present in my private packagist. 

TL;DR
The zendframework packages (actually no package at all are found by "findPackage") are not being returned in case that packagist.org is disabled as a repository plus a private packagist provides its package.json 
Satis does not provide "partial providers" informations and therefore no issue here. 


Long Story 
The thing is, that "getProviderNames" returns different Informations if the packages.json tells composer that it is splitted into partials. 
Ive tested around with our private packagist which returns the information that it has partial informations about the providers but still has a full list of packages it provides. 
Thus, I am not sure why there is a handling for "hasPartialProviders" in the first place.
 
Probably some legacy stuff? I found a work around without creating a new assumption of having "packages" available in the package.json

**Comment by Seldaek** [comment]
Ok merging then

**Comment by Seldaek** [comment]
thanks for the patch, I hope you can confirm it works with the latest composer.phar snapshot

**Comment by boesing** [comment]
Confirming that this patch solved the dependency resolution issue with laminas migration tool.
Thank you!
